core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -453,20 +459,47 @@ private void validateNoNewDeletesForDataFiles(
       // fail
       DeleteFile[] deleteFiles = deletes.forDataFile(startingSequenceNumber, dataFile);
       if (ignoreEqualityDeletes) {
-        ValidationException.check(
-            Arrays.stream(deleteFiles)
-                .noneMatch(deleteFile -> deleteFile.content() == FileContent.POSITION_DELETES),
-            "Cannot commit, found new position delete for replaced data file: %s",
-            dataFile);
+        if (!ignorePositionDeletes) {
+          ValidationException.check(
+              Arrays.stream(deleteFiles)
+                  .noneMatch(deleteFile -> deleteFile.content() == FileContent.POSITION_DELETES),
+              "Cannot commit, found new position delete for replaced data file: %s",
+              dataFile);
+        }
       } else {
-        ValidationException.check(
-            deleteFiles.length == 0,
-            "Cannot commit, found new delete for replaced data file: %s",
-            dataFile);
+        if (ignorePositionDeletes) {
+          ValidationException.check(
+              Arrays.stream(deleteFiles)
+                  .noneMatch(deleteFile -> deleteFile.content() == FileContent.EQUALITY_DELETES),
+              "Cannot commit, found new equality delete for replaced data file: %s",
+              dataFile);
+        } else {
+          ValidationException.check(
+              deleteFiles.length == 0,
+              "Cannot commit, found new delete for replaced data file: %s",
+              dataFile);
+        }
```

Comment:
Seems like we can rewrite the logic of L454-L481  to be like this:
```
if(ignoreEq && ignorePos) {
 return;
}

...

for (DataFile dataFile : dataFiles) {
  DeleteFile[] deleteFiles = deletes.forDataFile(startingSequenceNumber, dataFile);
  if(!ignoreEq) {
    validate that if any eq delete exists in deleteFiles
 }

  if(!ignorePos) {
    validate that if any pos delete exists in deleteFiles
  }
}
```
Does this cover the current validation logic ?

Reply:
That's a good point!